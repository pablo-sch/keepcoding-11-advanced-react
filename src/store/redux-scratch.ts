// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck: no check this file

/**
 * Este cÃ³digo crea un store personalizado que mantiene el estado,
 * permite despachar acciones para modificarlo mediante un reducer,
 * y permite a otras funciones suscribirse para ser notificadas de los cambios.
 */
export function createStore(reducer, initialState) {
  let state = initialState;
  let listeners = [];

  function getState() {
    return state;
  }

  function dispatch(action) {
    state = reducer(state, action);
    listeners.forEach((l) => l());
  }

  function subscribe(listener) {
    listeners.push(listener);

    return function unsubscribe() {
      listeners = listeners.filter((l) => l !== listener);
    };
  }

  dispatch({ type: "INIT" });

  return {
    getState,
    dispatch,
    subscribe,
  };
}
