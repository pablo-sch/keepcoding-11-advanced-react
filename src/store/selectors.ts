import type { RootState } from ".";

export const getIsLogged = (state: RootState) => state.auth;

export const getAdverts = (state: RootState) => state.adverts ?? [];

// export const getTweet = (state: RootState, advertId?: string) =>
//   state.adverts.find((advert) => advert.id === Number(advertId));

export function getAdvert(advertId?: string) {
  return function (state: RootState) {
    return state.adverts?.find((advert) => advert.id === advertId);
  };
}

export const getUi = (state: RootState) => state.ui;
