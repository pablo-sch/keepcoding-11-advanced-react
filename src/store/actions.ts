//DEPENDENCIES
import type { AppThunk } from ".";
import { login } from "../pages/auth/service";
import type { Credentials } from "../pages/auth/types";

//SERVICES
import { createAdvert, getAdverts, getAdvert as getAdvertService, deleteAdvert, getTags } from "../pages/advert/service";
import type { Advert } from "../pages/advert/types";
import { getAdvert } from "./selectors";

//Tipos de acciones===========================================================================================================
/**
 * Define los tipos de acciones que pueden ser enviadas (dispatch) al store,
 * incluyendo acciones para login, logout, adverts y errores de UI.
 */
type AuthLoginPending = {
  type: "auth/login/pending";
};

type AuthLoginFulfilled = {
  type: "auth/login/fulfilled";
};

type AuthLoginRejected = {
  type: "auth/login/rejected";
  payload: Error;
};

type AuthLogout = {
  type: "auth/logout";
};

// ................................................

type AdvertsLoadedFulFilled = {
  type: "adverts/loaded/fulfilled";
  payload: Advert[];
};

type AdvertsCreatedFulFilled = {
  type: "adverts/created/fulfilled";
  payload: Advert;
};

type AdvertsDetailFulFilled = {
  type: "adverts/detail/fulfilled";
  payload: Advert;
};

type AdvertsDeletedFulfilled = {
  type: "adverts/deleted/fulfilled";
  payload: string;
};

// ................................................

type TagsLoadedFulfilled = {
  type: "tags/loaded/fulfilled";
  payload: string[];
};

// ................................................

type UiResetError = {
  type: "ui/reset-error";
};

//Acciones sincrónicas============================================================================================================
/**
 * Funciones que devuelven objetos de acción con un type (y opcionalmente un payload).
 * Se usan para actualizar el estado del store.
 */
export const authLoginPending = (): AuthLoginPending => ({
  type: "auth/login/pending",
});

export const authLoginFulfilled = (): AuthLoginFulfilled => ({
  type: "auth/login/fulfilled",
});

export const authLoginRejected = (error: Error): AuthLoginRejected => ({
  type: "auth/login/rejected",
  payload: error,
});

//Thunks (acciones asíncronas)===============================================================================================0
/**
 * Funciones que permiten ejecutar lógica asíncrona (como llamadas a la API) y luego despachar acciones según el resultado.
 * Usan el middleware redux-thunk.
 */
export function authLogin(credentials: Credentials): AppThunk<Promise<void>> {
  return async function (dispatch) {
    dispatch(authLoginPending());
    try {
      await login(credentials);
      dispatch(authLoginFulfilled());
    } catch (error) {
      if (error instanceof Error) {
        dispatch(authLoginRejected(error));
      }
      throw error;
    }
  };
}

export const authLogout = (): AuthLogout => ({
  type: "auth/logout",
});

// ................................................

export const advertsLoadedFulFilled = (adverts: Advert[]): AdvertsLoadedFulFilled => ({
  type: "adverts/loaded/fulfilled",
  payload: adverts,
});

export const advertsDetailFulFilled = (advert: Advert): AdvertsDetailFulFilled => ({
  type: "adverts/detail/fulfilled",
  payload: advert,
});

export const advertsCreatedFulfilled = (advert: Advert): AdvertsCreatedFulFilled => ({
  type: "adverts/created/fulfilled",
  payload: advert,
});

export const advertsDeletedFulfilled = (advertId: string): AdvertsDeletedFulfilled => ({
  type: "adverts/deleted/fulfilled",
  payload: advertId,
});

export function advertsLoaded(): AppThunk<Promise<void>> {
  return async function (dispatch, getState) {
    const state = getState();
    if (state.adverts.loaded) {
      return;
    }
    try {
      const adverts = await getAdverts();
      dispatch(advertsLoadedFulFilled(adverts));
    } catch (error) {
      console.log(error);
    }
  };
}

// ................................................

export function advertsDelete(advertId: string): AppThunk<Promise<void>> {
  return async function (dispatch) {
    try {
      await deleteAdvert(advertId);
      dispatch(advertsDeletedFulfilled(advertId));
    } catch (error) {
      console.error("Error deleting advert:", error);
      throw error;
    }
  };
}

// ................................................

export function advertDetail(advertId: string): AppThunk<Promise<void>> {
  return async function (dispatch, getState) {
    const state = getState();
    if (getAdvert(advertId)(state)) {
      return;
    }
    try {
      const advert = await getAdvertService(advertId);
      dispatch(advertsDetailFulFilled(advert));
    } catch (error) {
      console.log(error);
      throw error;
    }
  };
}

export function advertsCreate(advertContent: FormData): AppThunk<Promise<Advert>> {
  return async function (dispatch) {
    try {
      const createdAdvert = await createAdvert(advertContent);
      const advert = await getAdvertService(createdAdvert.id.toString());
      dispatch(advertsCreatedFulfilled(advert));
      return advert;
    } catch (error) {
      console.log(error);
      throw error;
    }
  };
}

// ................................................

export function tagsLoaded(): AppThunk<Promise<void>> {
  return async function (dispatch) {
    try {
      const tags = await getTags();
      dispatch({ type: "tags/loaded/fulfilled", payload: tags });
    } catch (error) {
      console.error("Error loading tags:", error);
      throw error;
    }
  };
}

//Acción para resetear errores de UI============================================================================================
/**
 * Limpia errores en la interfaz, útil después de mostrar mensajes de error al usuario.
 */
export const uiResetError = (): UiResetError => ({
  type: "ui/reset-error",
});

//Unificacion de Acciones----------------------------------------------------------------
/**
 * Agrupa todos los tipos de acciones posibles en un solo tipo (Actions),
 * usado para tipar correctamente el dispatch y los reducers.
 */
export type Actions = AuthLoginPending | AuthLoginFulfilled | AuthLoginRejected | AuthLogout | AdvertsLoadedFulFilled | AdvertsDetailFulFilled | AdvertsCreatedFulFilled | AdvertsDeletedFulfilled | TagsLoadedFulfilled | UiResetError;
