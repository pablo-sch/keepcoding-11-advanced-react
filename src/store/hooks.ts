//REACT
import type { Credentials } from "../pages/auth/types";

//REDUX
import { authLogin, authLogout, uiResetError } from "./actions";
import { getIsLogged } from "./selectors";
import { useAppDispatch, useAppSelector } from ".";

export function useAuth() {
  return useAppSelector(getIsLogged);
}

export function useLoginAction() {
  const dispatch = useAppDispatch();
  return function (credentials: Credentials) {
    return dispatch(authLogin(credentials));
  };
}

export function useLogoutAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogout());
  };
}

export function useUiResetError() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(uiResetError());
  };
}
