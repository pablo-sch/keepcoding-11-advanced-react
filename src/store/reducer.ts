//REACT
import type { Advert } from "../pages/advert/types";

//REDUX
import { type Actions, type ActionsRejected } from "./actions";

export type State = {
  auth: boolean;
  adverts: {
    loaded: boolean;
    data: Advert[];
  };
  tags: string[];
  ui: {
    pending: boolean;
    error: Error | null;
  };
};

const defaultState: State = {
  auth: false,
  adverts: {
    loaded: false,
    data: [],
  },
  tags: [],
  ui: {
    pending: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action: Actions): State["auth"] {
  switch (action.type) {
    case "auth/login/fulfilled":
      return true;
    case "auth/logout":
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action: Actions): State["adverts"] {
  switch (action.type) {
    case "adverts/loaded/fulfilled":
      return { loaded: true, data: action.payload };
    case "adverts/detail/fulfilled":
      return { loaded: false, data: [action.payload] };
    case "adverts/created/fulfilled":
      return { ...state, data: [action.payload, ...(state.data ?? [])] };
    case "adverts/deleted/fulfilled":
      return {
        ...state,
        data: state.data.filter((advert) => advert.id !== action.payload),
      };
    default:
      return state;
  }
}

export function tags(state = defaultState.tags, action: Actions): State["tags"] {
  switch (action.type) {
    case "tags/loaded/fulfilled":
      return action.payload;
    default:
      return state;
  }
}

function isRejectedAction(action: Actions): action is ActionsRejected {
  return action.type.endsWith("/rejected");
}

export function ui(state = defaultState.ui, action: Actions): State["ui"] {
  if (action.type === "auth/login/pending") {
    return { pending: true, error: null };
  }
  if (action.type === "auth/login/fulfilled") {
    return { pending: false, error: null };
  }
  if (isRejectedAction(action)) {
    return { pending: false, error: action.payload };
  }
  if (action.type === "ui/reset-error") {
    return { ...state, error: null };
  }
  return state;
}
