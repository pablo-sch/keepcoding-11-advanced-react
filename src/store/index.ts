//DEPENDENCIES
import { combineReducers, createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import { useDispatch, useSelector } from "react-redux";
import type { createBrowserRouter } from "react-router";
import * as thunk from "redux-thunk";

//REACT
import * as adverts from "../pages/advert/service";
import * as auth from "../pages/auth/service";

//REDUX
import type { Actions } from "./actions";
import * as reducers from "./reducer";

// Combinación de reducers-------------------------------------------------------------------------------------------------------
const rootReducer = combineReducers(reducers);
type Router = ReturnType<typeof createBrowserRouter>;

export type ExtraArgument = {
  api: { auth: typeof auth; adverts: typeof adverts };
  router: Router;
};

// @ts-expect-error: any
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const timestamp = (store) => (next) => (action) => {
  const nextAction = {
    ...action,
    meta: {
      ...action.meta,
      timestamp: new Date(),
    },
  };
  return next(nextAction);
};

// @ts-expect-error: any
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const failureRedirects = (router: Router) => (store) => (next) => (action) => {
  const result = next(action);
  if (!action.type.endsWith("/rejected")) {
    return result;
  }

  if (action.payload.status === 404) {
    router.navigate("/not-found");
  }

  if (action.payload.status === 401) {
    router.navigate("/login");
  }

  if (action.payload.code === "ERR_NETWORK") {
    router.navigate("/internal-server-error");
  }
};

// Configuración del store-----------------------------------------------------------------------------------------------------
export default function configureStore(preloadedState: Partial<reducers.State>, router: Router) {
  const store = createStore(
    rootReducer,
    preloadedState as never,
    // // // @ts-expect-error: import devtools extension
    // // window.__REDUX_DEVTOOLS_EXTENSION__ &&
    // //   // @ts-expect-error: import devtools extension
    // //   window.__REDUX_DEVTOOLS_EXTENSION__(),
    composeWithDevTools(
      applyMiddleware(
        thunk.withExtraArgument<reducers.State, Actions, ExtraArgument>({
          api: { adverts, auth },
          router,
        }),
        timestamp,
        failureRedirects(router)
      )
    )
  );
  return store;
}

// Tipos para TypeScript-------------------------------------------------------------------------------------------------------
export type AppStore = ReturnType<typeof configureStore>;
export type AppGetState = AppStore["getState"];
export type RootState = ReturnType<AppGetState>;
export type AppDispatch = AppStore["dispatch"];

// Hooks personalizados y tipados----------------------------------------------------------------------------------------------
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();

// Tipo AppThunk para acciones asíncronas---------------------------------------------------------------------------------------
export type AppThunk<ReturnType = void> = thunk.ThunkAction<ReturnType, RootState, ExtraArgument, Actions>;
