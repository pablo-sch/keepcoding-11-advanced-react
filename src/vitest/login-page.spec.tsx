import { render, screen } from "@testing-library/react";
import LoginPage from "./../pages/auth/login-page";
import { Provider } from "react-redux";
import { authLogin, uiResetError } from "../store/actions";
import userEvent from "@testing-library/user-event";
import type { RootState } from "../store";

vitest.mock("../store/actions");

describe("LoginPage", () => {
  const state: RootState = {
    auth: false,
    adverts: {
      loaded: false,
      data: [],
    },
    tags: [],
    ui: {
      pending: false,
      error: null,
    },
  };
  const renderComponent = (error?: Error) => {
    if (error) {
      state.ui.error = error;
    }
    return render(
      <Provider
        store={{
          getState: () => state,
          //@ts-expect-error: subscribe
          subscribe: () => {},
          //@ts-expect-error: dispatch
          dispatch: () => {},
        }}
      >
        <LoginPage />
      </Provider>
    );
  };

  test("should render", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch login action on submit", async () => {
    renderComponent();

    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/Password/);
    const button = screen.getByRole("button");

    expect(button).toHaveTextContent("Log In");
    expect(button).toBeDisabled();

    //fireEvent.change(emailInput, { target: { value: "example@gmail.com" } });
    await userEvent.type(emailInput, "example@gmail.com");

    //fireEvent.change(passwordInput, { target: { value: "1234" } });
    await userEvent.type(passwordInput, "1234");

    expect(button).toBeEnabled();

    //fireEvent.click(button);
    await userEvent.click(button);

    expect(authLogin).toHaveBeenCalledWith({ email: "example@gmail.com", password: "1234" });
  });

  test("should render error", async () => {
    const error = new Error("Wrong gmail/password");
    const { container } = renderComponent(error);
    expect(container).toMatchSnapshot();

    const alert = screen.getByRole("alert");
    expect(alert).toHaveTextContent(error.message);

    await userEvent.click(alert);

    expect(uiResetError).toHaveBeenCalled();
  });
});
