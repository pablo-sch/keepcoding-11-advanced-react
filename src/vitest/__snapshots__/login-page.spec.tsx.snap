// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-gray-100 px-4 bg-gray-300"
  >
    <form
      class="max-w-xl mx-auto bg-gray-100 p-6 rounded-lg shadow space-y-6"
    >
      <h1
        class="text-2xl font-bold text-center text-gray-800"
      >
        Log in to your account
      </h1>
      <div
        class=""
      >
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="email"
        >
          Email
        </label>
        <input
          class="w-full px-3 py-2 rounded-md bg-gray-100 text-sm border border-gray-300 
            focus:outline-none focus:ring-2 focus:ring-blue-500 
            hover:cursor-text hover:border-gray-400 transition-colors"
          id="email"
          name="email"
          placeholder="e.g. user@example.com"
          type="email"
          value=""
        />
      </div>
      <div
        class=""
      >
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="password"
        >
          Password
        </label>
        <input
          class="w-full px-3 py-2 rounded-md bg-gray-100 text-sm border border-gray-300 
            focus:outline-none focus:ring-2 focus:ring-blue-500 
            hover:cursor-text hover:border-gray-400 transition-colors"
          id="password"
          name="password"
          placeholder="Your password"
          type="password"
          value=""
        />
      </div>
      <button
        class="sc-bRKDuR eYxMyf inline-flex items-center justify-center rounded-full px-6 py-2 font-semibold min-w-[72px] min-h-[36px] transition-colors duration-200 cursor-pointer bg-blue-500 text-white border border-blue-500 hover:bg-blue-600 opacity-50 pointer-events-none w-full"
        disabled=""
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginPage > should render error 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-gray-100 px-4 bg-gray-300"
  >
    <form
      class="max-w-xl mx-auto bg-gray-100 p-6 rounded-lg shadow space-y-6"
    >
      <h1
        class="text-2xl font-bold text-center text-gray-800"
      >
        Log in to your account
      </h1>
      <div
        class=""
      >
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="email"
        >
          Email
        </label>
        <input
          class="w-full px-3 py-2 rounded-md bg-gray-100 text-sm border border-gray-300 
            focus:outline-none focus:ring-2 focus:ring-blue-500 
            hover:cursor-text hover:border-gray-400 transition-colors"
          id="email"
          name="email"
          placeholder="e.g. user@example.com"
          type="email"
          value=""
        />
      </div>
      <div
        class=""
      >
        <label
          class="block text-sm font-medium text-gray-700 mb-1"
          for="password"
        >
          Password
        </label>
        <input
          class="w-full px-3 py-2 rounded-md bg-gray-100 text-sm border border-gray-300 
            focus:outline-none focus:ring-2 focus:ring-blue-500 
            hover:cursor-text hover:border-gray-400 transition-colors"
          id="password"
          name="password"
          placeholder="Your password"
          type="password"
          value=""
        />
      </div>
      <button
        class="sc-bRKDuR eYxMyf inline-flex items-center justify-center rounded-full px-6 py-2 font-semibold min-w-[72px] min-h-[36px] transition-colors duration-200 cursor-pointer bg-blue-500 text-white border border-blue-500 hover:bg-blue-600 opacity-50 pointer-events-none w-full"
        disabled=""
        type="submit"
      >
        Log In
      </button>
      <div
        aria-live="assertive"
        class="mt-4 p-3 bg-red-100 text-red-700 border border-red-300 rounded cursor-pointer text-sm text-center"
        role="alert"
      >
        ⚠️ Wrong gmail/password
      </div>
    </form>
  </div>
</div>
`;
